### PR TITLE
MM-14827 don't kill STDOUT for the server process

### PR DESCRIPTION
**Pull request by scottleedavis**
#### Summary
Prevent `make stop-server` from killing the logs process, and thus killing the server prematurely. We now both handle `SIGPIPE` and keep the pipe open going forward.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/10735

### CODE REVIEW

**Comment by levb** [comment]
I am 2/5 that we should try to find and fix the root cause - the processes shouldn't be left running to start with?

**Comment by cpoile** [comment]
I'm also partial to trying to find the root cause of this.

**Comment by scottleedavis** [comment]
I dug a bit deeper and it seems that [terminating a parent process on OSX does not necessarily  terminate it's child processes](https://superuser.com/questions/99789/when-a-python-process-is-killed-on-os-x-why-doesnt-it-kill-the-child-processes).

I tested between OSX Mojave and Linux Ubuntu 18.04 the following go snippet that spawns a child process.

```go
package main

import "os/exec"

func main() {
    lsCmd := exec.Command("bash", "-c", "sleep 1000")
    _, err := lsCmd.Output()
    if err != nil {
        panic(err)
    }
}
```

I then watched the process group, and manually terminated the parent process _using the pid_.
```bash
#finds the process group id
PGID=$(ps ax -O pgid | grep [go]-build |  awk '{printf $2}')

#shows the whole process group
ps -aux | grep $PGID
```

The child process would always persist on OSX, and always terminate on linux.  (sample size of 3 times each).


Given this finding I decided to update the linux block of `stop-server` to terminate processes based off of process group id instead of pid, which will terminate the entire group of parent and child.

**Comment by cpoile** [comment]
@scottleedavis That's great! So, as I understand it, OSX will still leave the children processes running?

**Comment by cpoile** [comment]
@scottleedavis For OSX: Do you think you could just gather the pids of the pgid of the go process, and then kill them all that way?

**Comment by levb** [comment]
2/5 we should be attempting to gracefully shut down plugin processes (TBD exactly what this means) on mattermost-server shutdown. Which means we should keep track of them, and can kill them explicitly?

**Comment by scottleedavis** [comment]
@cpoile While that is possible, I chose to kill the entire process group as that was less lines of code vs. finding the group and loop through each pid.  Definitely can do what you request though.

@levb I will investigate more on your idea.  I interpreted it to mean 'somewhere in server code, capture the termination signal and explicitly terminate plugins'.

**Comment by cpoile** [comment]
@scottleedavis I agree with Lev, trying to track the ids and destroying them explicitly would be ideal. But, if worse comes to worse, this works as a last ditch effort. Run in the mattermost-server/ directory, iterate through each plugin directory (which will be in the name of the plugin process) and kill that process:
https://community.mattermost.com/core/pl/jmdie1u7sjnwxeaboq6azgogua

**Comment by scottleedavis** [comment]
@cpoile @levb the latest commit reverts the Makefile changes and instead changed signal.Notify in runServer.  This signal was not being triggered with the specified syscall constraints on OSX.  I tried every known syscall one by one and it did not change the results.  
When I used 'all' syntax (e.g. signal.Notify(c)), the `make stop-server`calls triggered the signal and the ShutdownServer function was triggered, with the end result of the plugins being deactivated gracefully.  

However, it seems signal.Notify emits strange unsightly output to stop-server now.
```
$ make stop-server
Stopping mattermost
stopping go 73342
stopping go 73343
stopping mattermost 73391
stopping mattermost 73392
2019-05-15 22:20:08.327586 -0700 PDT m=+3.128610708 write error: write /dev/stdout: broken pipe
2019-05-15 22:20:08.327694 -0700 PDT m=+3.128718762 write error: write /dev/stdout: broken pipe
2019-05-15 22:20:08.32777 -0700 PDT m=+3.128794617 write error: write /dev/stdout: broken pipe
2019-05-15 22:20:08.327811 -0700 PDT m=+3.128835651 write error: write /dev/stdout: broken pipe
2019-05-15 22:20:08.327804 -0700 PDT m=+3.128829612 write error: write /dev/stdout: broken pipe
2019-05-15 22:20:08.327944 -0700 PDT m=+3.128968960 write error: write /dev/stdout: broken pipe
2019-05-15 22:20:08.328301 -0700 PDT m=+3.129325595 write error: write /dev/stdout: broken pipe
```

**Comment by scottleedavis** [comment]
I was able to white it down to the SIGPIPE syscall that enables the signal catch.
`signal.Notify(interruptChan, os.Interrupt, syscall.SIGINT, syscall.SIGTERM, syscall.SIGPIPE)`

Although the plugins have shutdown, the `/dev/stdout: broken pipe` is still happening.  I will continue to look into this.

**Comment by scottleedavis** [comment]
@levb @cpoile 

I feel the SIGPIPE approach was incorrect, so I looked deeper into how plugins are supposed to be shutting down.  It turns out that the plugin supervisor is already tracking the pid for the plugin process.  Using that, I was able to kill the process on shutdown, however... only when the graceful approach was not done (i.e. OnDeactivate & client.Kill).  For example, the following looks near ideal to me.  (I added the syscall.Kill)
```go
// Shutdown deactivates all plugins and gracefully shuts down the environment.
func (env *Environment) Shutdown() {
	env.activePlugins.Range(func(key, value interface{}) bool {
		ap := value.(activePlugin)

		if ap.supervisor != nil {
			if err := ap.supervisor.Hooks().OnDeactivate(); err != nil {
				env.logger.Error("Plugin OnDeactivate() error", mlog.String("plugin_id", ap.BundleInfo.Manifest.Id), mlog.Err(err))
			}
			ap.supervisor.Shutdown()
			syscall.Kill(ap.supervisor.pid, syscall.SIGHUP)
		}

		env.activePlugins.Delete(key)

		return true
	})
}
```
unfortunately, if either OnDeactivate, or Shutdown are left in, the plugin process does not kill.  However when syscall.Kill is used alone, the plugin processes kill every time.  (much less graceful)

Both functions rely on the supervisor client (hashicorp go-plugin), so more digging would seem to be needed inside that import.


Would you want me to dig more into the hashicorp go-plugin to see if a fix can be made there?
Do you think the less graceful approach (no ondeactivate or client.Kill) is accetable?   (i currently don't)

**Comment by scottleedavis** [comment]
:( although from the build log, apparently `syscall.Kill` isn't working on windows
```
plugin/environment.go:305:4: undefined: syscall.Kill
build/release.mk:13: recipe for target 'build-windows' failed
make: *** [build-windows] Error 2
```

**Comment by levb** [comment]
@scottleedavis looking good! I'll test this today, also want refresh my memory on proper use of signals in Unix before I confirm.

**Comment by cpoile** [comment]
@levb gentle reminder about this -- I'll defer to your understanding of signals.

@scottleedavis Very nice find! Are you able to test on windows and linux? Does removing the old `Hooks().OnDeactivate()` and `supervisor.Shupdown()` cause problems on those OSes?

**Comment by scottleedavis** [comment]
@cpoile It will take a bit but I'd be happy to test those OSes as well.  👍

File: plugin/environment.go [hunk 1]
```diff
@@ -300,10 +301,14 @@ func (env *Environment) Shutdown() {
 		ap := value.(activePlugin)
 
 		if ap.supervisor != nil {
-			if err := ap.supervisor.Hooks().OnDeactivate(); err != nil {
-				env.logger.Error("Plugin OnDeactivate() error", mlog.String("plugin_id", ap.BundleInfo.Manifest.Id), mlog.Err(err))
+			//if err := ap.supervisor.Hooks().OnDeactivate(); err != nil {
+			//	env.logger.Error("Plugin OnDeactivate() error", mlog.String("plugin_id", ap.BundleInfo.Manifest.Id), mlog.Err(err))
+			//}
+			//ap.supervisor.Shutdown()
```
Comment (levb) [inline plugin/environment.go]: Why is ap.supervisor.Shutdown() commented out?

File: plugin/environment.go [hunk 2]
```diff
@@ -300,10 +301,14 @@ func (env *Environment) Shutdown() {
...
+			p, err := os.FindProcess(ap.supervisor.pid)
```
Comment (levb) [inline plugin/environment.go]: I don't understand the exact role of the supervisor, but as a general rule we should shut things down gracefully. 1/5 The supervisor then should attempt to kill the plugins, however it decides fit (including just exiting if it works).

File: plugin/environment.go [hunk 3]
```diff
@@ -300,10 +301,14 @@ func (env *Environment) Shutdown() {
...
+			if err == nil {
+				p.Signal(syscall.SIGHUP)
```
Comment (levb) [inline plugin/environment.go]: SIGHUP in servers is usually used these days to (gracefully) reload the config. See https://stackoverflow.com/a/690631 for a pretty good explanation.

To be clear, I think in this case we should be sending a SIGTERM first (maybe in Shutdown?), and then if it times out (i.e. the process is still there) send a SIGKILL.

File: plugin/environment.go (hunk 1 above)
Comment (scottleedavis) [reply plugin/environment.go]: @levb it was left it to show that I would prefer it there...  However if either `			//ap.supervisor.Shutdown()`, or `ap.supervisor.Hooks().OnDeactivate()` is left in, the `p.Signal` call doesn't seem to fire or effectively kill the plugin process.

File: plugin/environment.go (hunk 2 above)
Comment (scottleedavis) [reply plugin/environment.go]: I would prefer it in too.   however, the `ap.supervisor.Shutdown` process calls `client.Kill()`, which is the hashicorp go-plugin client code, and seems to conflict with the direct `p.Signal` call.   (see above on the combined issue canceling out the pid signal)

Comment (scottleedavis) [reply plugin/environment.go]: @levb I can look further into the hashicorp go-plugin code (perhaps we can find a bug there), though we are on their latest release https://github.com/hashicorp/go-plugin/releases

File: plugin/environment.go (hunk 3 above)
Comment (scottleedavis) [reply plugin/environment.go]: Added SIGTERM, and if error returned SIGKILL.  Would you prefer a time out check vs. error check?

**Comment by scottleedavis** [comment]
@cpoile on windows 10 `make stop-server` correctly shuts down with these changes.   I'll review linux next.

**Comment by scottleedavis** [comment]
@cpoile ubuntu 18.04 `make stop-server` correctly shuts down with these changes too.

**Comment by scottleedavis** [comment]
@levb As it is preferable to let the supervisor manage the shutdown/kill of plugins, do you feel:
1) we should revert to using the Makefile to kill orphaned plugins
2) look further into hashicorps go-plugin client on osx (why isn't the process killing?)
3) Follow the approach we have hashed out so far, (though honestly it feels less than elegant)
4) other?

**Comment by cpoile** [comment]
@scottleedavis That's great work Scott, thanks for working on this. I'll wait for Lev's comments about what would be the preferred path forward.

**Comment by cpoile** [comment]
@levb just a nudge about this, now that jira v2.0 is done. :)

**Comment by lieut-data** [comment]
@scottleedavis, I've been asked to review this but am just catching up-to-speed.

Independent of OSX-specific issues, I do think we should materialize a strategy for dealing with a plugin that refuses to stop. Not sure what all has been tried, but I might suggest writing a "plugin-wont-die" that won't gracefully exit (not maliciously, but maybe just by a deadlock in `main` or some such).

Assuming that reproduces the hung experience, then I'd like to see us `SIGTERM` and then `SIGKILL` after 10 seconds, logging all of the actions being taken. Ideally this is all done in parallel if there are multiple hung plugins.

Does that jive with your thoughts on this issue?

**Comment by scottleedavis** [comment]
@lieut-data the crux of the OSX issue seems to be that when either `rp.supervisor.Hooks().OnDeactivate()` or `rp.supervisor.Shutdown()` is used, any subsequent `p.Signal(..)` doesn't effectively kill the plugin process.  (`p.Signal(...)` used by itself does however)

As far as writing a "plugin-wont-die" and the implementing the SIGTERM/SIGKILL approach I think that sounds good and will give that a try this weekend.  https://github.com/scottleedavis/mattermost-plugin-wont-die

**Comment by scottleedavis** [comment]
The [plugin-won't-die](https://github.com/scottleedavis/mattermost-plugin-wont-die) plugin deadlocks in main, and can be killed (but not gracefully) with a manual `disable` plugin action.  We may need to investigate a more robust _won't die_.

Updated this MR to include a `StopPlugin` function that checks if a plugin process is still active, then SIGTERM, waits 10 seconds and SIGKILL if plugin still active.  It logs every step as requested.

Since my dev station is OSX `StopPlugin` works by itself, but not with `OnDeactivate()` or `Shutdown()` as mentioned prior.

I have not yet dug into the hashicorp go-plugin client code to see where and what may be going on for at least OSX.  Thoughts @lieut-data ?

**Comment by lieut-data** [comment]
Thanks, @scottleedavis! Want to drill into one comment:
> @lieut-data the crux of the OSX issue seems to be that when either rp.supervisor.Hooks().OnDeactivate() or rp.supervisor.Shutdown() is used, any subsequent p.Signal(..) doesn't effectively kill the plugin process. (p.Signal(...) used by itself does however)

I'm not able to follow this description, in part because both `rp.supervisor.Hooks().OnDeactivate()` and `rp.supervisor.Shutdown()` are called right before the new `StopPlugin` function, but yet the contained `p.Signal` still works? Or is it that even this `StopPlugin` leaves a plugin process for you on OSX?

**Comment by scottleedavis** [comment]
@lieut-data What I mean is that on OSX, if I comment out those two functions, StopPlugin by itself kills the plugins.
```go
			//if err := rp.supervisor.Hooks().OnDeactivate(); err != nil {
			//	env.logger.Error("Plugin OnDeactivate() error", mlog.String("plugin_id", rp.BundleInfo.Manifest.Id), mlog.Err(err))
			//}
			//rp.supervisor.Shutdown()
			env.StopPlugin(rp.supervisor.pid)
```

However if I run them all uncommented together, the plugin processes persist after the server stopping.

I am not clear the _why_ behind this, though I have suspected something with the hashicorp go-plugin.  I am very open to being wrong and finding/fixing a root cause.

**Comment by scottleedavis** [comment]
The original approach of killing the plugins directly in the Makefile seems appealing vs. a code block that isn't effective (yet).  Perhaps I should dig more into the hashicorp plugin code?

**Comment by hanzei** [comment]
@lieut-data Can you please provide some guidance for @scottleedavis?

**Comment by hanzei** [comment]
@lieut-data genlte reminder about this topic

**Comment by lieut-data** [comment]
Thanks, @hanzei! @scottleedavis and I synced up offline, and I'm going to find some mana in the coming sprint(s) to help the investigation here.

**Comment by hanzei** [comment]
Thanks for the update @lieut-data . I just wanted to make sure this PR doesn't get forgotten.

**Comment by lieut-data** [comment]
@scottleedavis, thanks for the feedback.

1. Currently, with the `SIGPIPE` changes, I no longer see the server failing to try to stop the plugin processes. If nothing else goes wrong, this results in 100% cleanup. However, a few other things could go wrong, which are fixed in https://github.com/mattermost/mattermost-server/pull/11972 and https://github.com/mattermost/mattermost-server/pull/11988. Can you try out the combination of all these PRs to see if your observations change?

2. Fixing the `Makefile` is a solid change, since even with the core issue resolved, it's not ideal to kill STDOUT for the server process. 

> Is resorting to manually killing plugins in the make stop-server portion of the Makefile equivalent to not addressing the root cause? Very likely, though would be a stop gap.

Agreed that the `Makefile` changes alone don't address the root cause -- but I'm pretty convinced that the `SIGPIPE` issue /is/ the root cause, and likely any remaining issues are caught by the linked PRs above. Let me know your thoughts!

**Comment by scottleedavis** [comment]
@lieut-data with merged changes from [Stop plugin healthchecks](https://github.com/mattermost/mattermost-server/pull/11972) (already in master) and [Improved OnDeactivate handling](https://github.com/mattermost/mattermost-server/pull/11988) I confirm that the plugin processes do die 100% of the time.  :) 👍 

However on my system I am still seeing a slew of
```bash
signal: terminated
2019-08-29 16:24:13.464483 -0700 PDT m=+38.085069723 write error: write /dev/stdout: broken pipe
...
```
after `make stop(-server)`

Is there something different in my setup that I may have incorrect?  
I am running go 1.12.5 on osx 10.14.6, with bash 3.2.57(1)-release.

**Comment by scottleedavis** [comment]
branch with all merged branches https://github.com/scottleedavis/mattermost-server/tree/test-MM-14827

**Comment by lieut-data** [comment]
@scottleedavis, it looks like your fixes here to `make stop-server` are still killing the `mail logs --logrus` process, so the server still ends up with no where to which to pipe `STDOUT`.

If you change extend the check to include the second `kill`, it should work as expected:
```go
	@for PID in $$(ps -ef | grep "[g]o run" | grep "disableconfigwatch" | awk '{ print $$2 }'); do \
		echo stopping go $$PID; \
		kill $$PID; \
	done
	@for PID in $$(ps -ef | grep "[g]o-build" | grep "disableconfigwatch" | awk '{ print $$2 }'); do \
		echo stopping mattermost $$PID; \
		kill $$PID; \
	done
```

**Comment by scottleedavis** [comment]
That works great @lieut-data , thank you.  pushed, and up to date with master.  Seeing the log output of the plugins being deactivated now is pretty neat.  ⚡️ 

Also, thank you @levb !  This is definitely a personal record for total bytes changed/net bytes changed.  😄

**Comment by lieut-data** [comment]
@prapti, to test this, it mostly suffices to setup a few plugins, invoke `make stop-server` and verify that the plugins all stop successfully and the server finishes normally. We may want to spot check other tests that relate to starting/stopping the server, as well as plugin activation.

**Comment by cpoile** [comment]
Thanks for pushing through on this @scottleedavis, and thanks @lieut-data for guiding it! Nice clean fix for the root cause. :)

**Comment by hanzei** [comment]
Thanks you very much for all the work you put into this @scottleedavis and @lieut-data :tada:

**Pull request merged**